src/lib/gadgets/rot.ts

```diff
@@ -0,0 +1,121 @@
+import { Field } from '../field.js';
+import { Provable } from '../provable.js';
+import { Fp } from '../../bindings/crypto/finite_field.js';
+import * as Gates from '../gates.js';
+
+export { rot, rotate };
+
+const MAX_BITS = 64 as const;
+
+function rot(word: Field, bits: number, direction: 'left' | 'right' = 'left') {
+  // Check that the rotation bits are in range
+  if (bits < 0 || bits > MAX_BITS) {
+    throw Error(`rot: expected bits to be between 0 and 64, got ${bits}`);
+  }
+
+  if (direction !== 'left' && direction !== 'right') {
+    throw Error(
+      `rot: expected direction to be 'left' or 'right', got ${direction}`
+    );
+  }
+
+  if (word.isConstant()) {
+    return new Field(Fp.rot(word.toBigInt(), bits, direction === 'left'));
+  }
+  const [rotated] = rotate(word, bits, direction);
+  return rotated;
+}
+
+function rotate(
+  word: Field,
+  bits: number,
+  direction: 'left' | 'right' = 'left'
+): [Field, Field, Field] {
+  // Check as the prover, that the input word is at most 64 bits.
+  Provable.asProver(() => {
+    if (word.toBigInt() > 2 ** MAX_BITS) {
+      throw Error(
+        `rot: expected word to be at most 64 bits, got ${word.toBigInt()}`
+      );
+    }
+  });
+
+  const rotationBits = direction === 'right' ? MAX_BITS - bits : bits;
+  const big2Power64 = 2n ** BigInt(MAX_BITS);
+  const big2PowerRot = 2n ** BigInt(rotationBits);
+
+  const [rotated, excess, shifted, bound] = Provable.witness(
+    Provable.Array(Field, 4),
+    () => {
+      const wordBigInt = word.toBigInt();
+
+      // Obtain rotated output, excess, and shifted for the equation:
+      // word * 2^rot = excess * 2^64 + shifted
+      const { quotient: excess, remainder: shifted } = divideWithRemainder(
+        wordBigInt * big2PowerRot,
+        big2Power64
+      );
+
+      // Compute rotated value as:
+      // rotated = excess + shifted
+      const rotated = shifted + excess;
+      // Compute bound that is the right input of FFAdd equation
+      const bound = excess + big2Power64 - big2PowerRot;
+      return [rotated, excess, shifted, bound].map(Field.from);
+    }
+  );
+
+  // Compute current row
+  Gates.rot(
+    word,
+    rotated,
+    excess,
+    [
+      witnessSlices(bound, 52, 64),
+      witnessSlices(bound, 40, 52),
+      witnessSlices(bound, 28, 40),
+      witnessSlices(bound, 16, 28),
+    ],
+    [
+      witnessSlices(bound, 14, 16),
+      witnessSlices(bound, 12, 14),
+      witnessSlices(bound, 10, 12),
+      witnessSlices(bound, 8, 10),
+      witnessSlices(bound, 6, 8),
+      witnessSlices(bound, 4, 6),
+      witnessSlices(bound, 2, 4),
+      witnessSlices(bound, 0, 2),
+    ],
+    Field.from(big2PowerRot)
+  );
+  // Compute next row
+  Gates.rangeCheck64(shifted);
+  // Compute following row
+  Gates.rangeCheck64(excess);
+  Gates.rangeCheck64(rotated);
+  return [rotated, excess, shifted];
+}
+
+function witnessSlices(f: Field, start: number, stop = -1) {
```

Comment:
One thing for the reviewer: I tried using t[his implementation for this function instead](https://github.com/o1-labs/o1js/pull/1177/files#diff-549b20f5f9b5702170289d8f7fe81901e79f801c2e07cfb35d1c9bf3b01a35e1R115), but I get issues with unit tests. Working on a solution to make this not an error but address the original comments.

Specifically I run into the following stacktrace:
```sh
Error: the lookup failed to find a match in the table: row=12
    at module.exports.__wbindgen_error_new (/home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/bindings/compiled/_node_bindings/plonk_wasm.cjs:9293:17)
    at <wasm_bindgen::JsError as core::convert::From<E>>::from::h745e731c0bba10da (wasm://wasm/01112d6e:wasm-function[4037]:0x3bd9f8)
    at caml_pasta_fp_plonk_proof_create (wasm://wasm/01112d6e:wasm-function[718]:0x24d985)
    at module.exports.caml_pasta_fp_plonk_proof_create (/home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/bindings/compiled/_node_bindings/plonk_wasm.cjs:2756:14)
    at async withThreadPool (file:///home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/bindings/js/node/node-backend.js:47:18)
    at async prettifyStacktracePromise (file:///home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/lib/errors.js:104:16)
    at async prove_ (file:///home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/lib/proof_system.js:165:26)
    at async file:///home/martin/Code/o1/mina/src/lib/snarkyjs/dist/node/lib/gadgets/gadgets.unit-test.js:56:17
``

Reply:
The function from the other PR has the `length` as input, rather than an `end` index - maybe that was the reason and the specified parameters exceeded that? Curious to see what it fails, can you point me to a PR?

Reply:
Ah, great point, I missed that :sweat: Fixed it and it passes the failing tests I had